explorer/common/arena.h

```diff
@@ -61,35 +119,162 @@ class Arena {
 
   // Templated destruction of a pointer.
   template <typename T>
-  class ArenaEntryTyped : public ArenaEntry {
-   public:
-    struct WriteAddressHelper {};
+  class ArenaEntryTyped;
 
-    template <typename... Args>
-    explicit ArenaEntryTyped(Args&&... args)
-        : instance_(std::forward<Args>(args)...) {}
+  // Manages allocations in an arena for destruction at shutdown.
+  std::vector<std::unique_ptr<ArenaEntry>> arena_;
+  int64_t allocated_ = 0;
 
-    template <typename... Args>
-    explicit ArenaEntryTyped(WriteAddressHelper, Args&&... args)
-        : ArenaEntryTyped(std::forward<Args>(args)...) {}
+  // Hash functor implemented in terms of hash_value (see llvm/ADT/Hashing.h)
+  struct LlvmHasher {
+    template <typename T>
+    auto operator()(const T& t) const -> size_t {
+      using llvm::hash_value;
+      return hash_value(t);
+    }
+  };
 
-    template <typename U, typename... Args>
-    explicit ArenaEntryTyped(WriteAddressTo<U> write_address, Args&&... args)
-        : ArenaEntryTyped(
-              (*write_address.target = &instance_, WriteAddressHelper{}),
-              std::forward<Args>(args)...) {}
+  // Factory metafunction for globally unique type IDs.
+  // &TypeId<T>::id == &TypeId<U>::id if and only if std::is_same_v<T,U>.
+  //
+  // Inspired by llvm::Any::TypeId.
+  template <typename T>
+  struct TypeId {
+    static char id;
```

Comment:
```suggestion
    // This is only used for an address to compare; the value is unimportant.
    static char id;
```

This had been stumping me what was significant about `1`, can you add a brief explanatory comment?